src/Lean/Compiler/IR/LLVMBindings.lean

```diff
@@ -0,0 +1,305 @@
+
+@[extern "lean_llvm_const_int"]
+opaque constInt (intty: @&Ptr LLVMType) (value: @&UInt64) (signExtend: @Bool := false): IO (Ptr Value)
+
+
+@[extern "lean_llvm_print_module_to_string"]
+opaque printModuletoString (mod: @&Ptr Module): IO (String)
+
+@[extern "lean_llvm_print_module_to_file"]
+opaque printModuletoFile (mod: @&Ptr Module) (file: @&String): IO Unit
+
+@[extern "llvm_count_params"]
+opaque countParams (fn: @&Ptr Function): UInt64 -- will this cause problems..?
+
+@[extern "llvm_get_param"]
+opaque getParam (fn: @&Ptr Function) (ix: UInt64): IO (Ptr Value)
+
+@[extern "lean_llvm_create_memory_buffer_with_contents_of_file"]
+opaque createMemoryBufferWithContentsOfFile (path: @&String): IO (Ptr MemoryBuffer)
+
+@[extern "lean_llvm_parse_bitcode"]
+opaque parseBitcode (context: @&Ptr Context) (membuf: @&Ptr MemoryBuffer): IO (Ptr Module)
+
+@[extern "lean_llvm_link_modules"]
+opaque linkModules (dest: @Ptr Module) (src: @&Ptr Module): IO Unit
+
+@[extern "lean_llvm_get_default_target_triple"]
+opaque getDefaultTargetTriple: IO String
+
+@[extern "lean_llvm_get_target_from_triple"]
+opaque getTargetFromTriple (triple: @&String): IO (Ptr Target)
+
+@[extern "lean_llvm_create_target_machine"]
+opaque createTargetMachine (target: @&Ptr Target) (tripleStr: @&String) (cpu: @&String) (features: @&String): IO (Ptr TargetMachine)
+
+@[extern "lean_llvm_target_machine_emit_to_file"]
+opaque targetMachineEmitToFile (targetMachine: @&Ptr TargetMachine) (module: @&Ptr Module) (filepath: @&String) (codegenType: @&UInt64): IO Unit
+
+
+-- Helper to add a function if it does not exist, and to return the function handle if it does.
+def getOrAddFunction(m: Ptr Module) (name: String) (type: Ptr LLVMType): IO (Ptr Value) :=  do
+  match (← getNamedFunction m name) with
+  | .some fn => return fn
+  | .none => addFunction m name type
+
+def getOrAddGlobal(m: Ptr Module) (name: String) (type: Ptr LLVMType): IO (Ptr Value) :=  do
+  match (← getNamedGlobal m name) with
+  | .some fn => return fn
+  | .none => addGlobal m name type
+
+
+def i1Type (ctx: LLVM.Ptr LLVM.Context): IO (LLVM.Ptr LLVM.LLVMType) :=
+  LLVM.intTypeInContext ctx 1
+
+def i8Type (ctx: LLVM.Ptr LLVM.Context): IO (LLVM.Ptr LLVM.LLVMType) :=
+  LLVM.intTypeInContext ctx 8
+
+def i16Type (ctx: LLVM.Ptr LLVM.Context): IO (LLVM.Ptr LLVM.LLVMType) :=
+  LLVM.intTypeInContext ctx 16
+
+def i32Type (ctx: LLVM.Ptr LLVM.Context): IO (LLVM.Ptr LLVM.LLVMType) :=
+  LLVM.intTypeInContext ctx 32
+
+
+def i64Type (ctx: LLVM.Ptr LLVM.Context): IO (LLVM.Ptr LLVM.LLVMType) :=
+  LLVM.intTypeInContext ctx 64
+
+def voidPtrType (ctx: LLVM.Ptr LLVM.Context): IO (LLVM.Ptr LLVM.LLVMType) :=
+  do LLVM.pointerType (← LLVM.intTypeInContext ctx 8)
+
+def i8PtrType (ctx: LLVM.Ptr LLVM.Context): IO (LLVM.Ptr LLVM.LLVMType) := voidPtrType ctx
+
+-- TODO: instantiate target triple and find out what size_t is.
+def size_tType (ctx: @&Ptr Context): IO (Ptr LLVMType) := i64Type ctx
```

Comment:
Why exactly is this hard-coded as 64 bit? Llvm does not even have a size t type. I am surprised the api offers such magic type interface. I would maybe drop this for now and add it when actually needed. I personally would probably put this type selection further out in the backend or maybe just call it differently. This functionality really is more about the c implementation on a given platform than about llvm.

Reply:
Fixed, this is removed from `LLVMBindings.lean`.